### PR TITLE
fix(files): prefer subdomain gw in copied share links

### PR DESCRIPTION
**Pull request by acul71**
## Issue #2244 
When you right click on a file in Files and choose "Share Link" you are given a URL like ```https://<host>/ipfs/<cid>```. This should be of the form ```https://<cid>.ipfs.<host>``` for security reasons.

## Solution
Modify the setting page and relevant code so that the user can adjust both subdomain and path gateway they use, but by default we use subdomain whenever possible

## Code

- Modified ```getShareableLink``` function in ```src/lib/files.js``` it now 
Generates a shareable link for the provided files using a subdomain gateway as default or a path gateway as fallback

```javascript
/**
 * Generates a shareable link for the provided files using a subdomain gateway as default or a path gateway as fallback.
 *
 * @param {FileStat[]} files - An array of file objects with their respective CIDs and names.
 * @param {string} gatewayUrl - The URL of the default IPFS gateway.
 * @param {string} subdomainGatewayUrl - The URL of the subdomain gateway.
 * @param {IPFSService} ipfs - The IPFS service instance for interacting with the IPFS network.
 * @returns {Promise<string>} - A promise that resolves to the shareable link for the provided files.
 */
export async function getShareableLink (files, gatewayUrl, subdomainGatewayUrl, ipfs) {
  let cid
  let filename

  if (files.length === 1) {
    cid = files[0].cid
    if (files[0].type === 'file') {
      filename = `?filename=${encodeURIComponent(files[0].name)}`
    }
  } else {
    cid = await makeCIDFromFiles(files, ipfs)
  }

  const url = new URL(subdomainGatewayUrl)

  /**
   * dweb.link (subdomain isolation) is listed first as the new default option.
   * However, ipfs.io (path gateway fallback) is also listed for CIDs that cannot be represented in a 63-character DNS label.
   * This allows users to customize both the subdomain and path gateway they use, with the subdomain gateway being used by default whenever possible.
   */
  let shareableLink = ''
  const base32Cid = cid.toV1().toString()
  if (base32Cid.length < 64) {
    shareableLink = `${url.protocol}//${base32Cid}.ipfs.${url.host}${filename || ''}`
  } else {
    shareableLink = `${gatewayUrl}/ipfs/${cid}${filename || ''}`
  }

  // console.log('Shareable link', shareableLink)
  return shareableLink
}
```

- Added in ```src/bundles/gateway.js``` a New function ```checkSubdomainGateway (gatewayUrl)``` that Checks if a given gateway URL is functioning correctly by verifying image loading and redirection. 

```javascript
/**
 * Checks if a given URL redirects to a subdomain that starts with a specific hash.
 *
 * @param {URL} url - The URL to check for redirection.
 * @throws {Error} Throws an error if the URL does not redirect to the expected subdomain.
 * @returns {Promise<void>} A promise that resolves if the URL redirects correctly, otherwise it throws an error.
 */
async function expectSubdomainRedirect (url) {
  // Detecting redirects on remote Origins is extra tricky,
  // but we seem to be able to access xhr.responseURL which is enough to see
  // if paths are redirected to subdomains.

  const { redirected, url: responseUrl } = await fetch(url.toString())
  console.log('redirected: ', redirected)
  console.log('responseUrl: ', responseUrl)

  if (redirected) {
    console.log('definitely redirected')
  }
  const { hostname } = new URL(responseUrl)

  if (!hostname.startsWith(IMG_HASH_1PX)) {
    const msg = `Expected ${url.toString()} to redirect to subdomain '${IMG_HASH_1PX}' but instead received '${responseUrl}'`
    console.log(msg)
    throw new Error(msg)
  }
}

/**
 * Checks if an image can be loaded from a given URL within a specified timeout.
 *
 * @param {URL} imgUrl - The URL of the image to be loaded.
 * @returns {Promise<void>} A promise that resolves if the image loads successfully within the timeout, otherwise it rejects with an error.
 */
async function checkViaImgUrl (imgUrl) {
  const imgCheckTimeout = 15000
  await new Promise((resolve, reject) => {
    const img = new Image()
    const timer = setTimeout(() => {
      reject(new Error(`Timeout when attempting to load img from '${img.src}`))
    }, imgCheckTimeout)
    img.onerror = () => {
      clearTimeout(timer)
      reject(new Error(`Error when attempting to load img from '${img.src}`))
    }
    img.onload = () => {
      clearTimeout(timer)
      console.log(`Successfully loaded img from '${img.src}'`)
      resolve(undefined)
    }
    img.src = imgUrl.toString()
  })
}

/**
 * Checks if a given gateway URL is functioning correctly by verifying image loading and redirection.
 *
 * @param {string} gatewayUrl - The URL of the gateway to be checked.
 * @returns {Promise<boolean>} A promise that resolves to true if the gateway is functioning correctly, otherwise false.
 */
export async function checkSubdomainGateway (gatewayUrl) {
  let imgSubdomainUrl
  let imgRedirectedPathUrl
  try {
    const gwUrl = new URL(gatewayUrl)
    imgSubdomainUrl = new URL(`${gwUrl.protocol}//${IMG_HASH_1PX}.ipfs.${gwUrl.hostname}/?now=${Date.now()}&filename=1x1.png#x-ipfs-companion-no-redirect`)
    imgRedirectedPathUrl = new URL(`${gwUrl.protocol}//${gwUrl.hostname}/ipfs/${IMG_HASH_1PX}?now=${Date.now()}&filename=1x1.png#x-ipfs-companion-no-redirect`)
  } catch (err) {
    console.log('Invalid URL:', err)
    return false
  }
  return await checkViaImgUrl(imgSubdomainUrl)
    .then(async () => expectSubdomainRedirect(imgRedirectedPathUrl))
    .then(() => {
      console.log(`Gateway at '${gatewayUrl}' is functioning correctly (verified image loading and redirection)`)
      return true
    })
    .catch((err) => {
      console.log(err)
      return false
    })
}
```

- Modified ui setting in ```settings/SettingsPage.js``` and labels in ```public/locales/en/app.json``` and ```public/locales/en/settings.json```
image

```javascript
     <Box className='mb3 pa4-l pa2'>
      <div className='lh-copy charcoal'>
      <Title>{t('app:terms.publicGateway')}</Title>
        <Trans i18nKey='publicSubdomainGatewayDescription' t={t}>
          <p>Choose which <a className='link blue' href="https://docs.ipfs.tech/concepts/ipfs-gateway/#subdomain" target='_blank' rel='noopener noreferrer'>Subdomain Gateway</a> you want to use when generating shareable links (Default: Subdomain Isolation)</p>
        </Trans>
        <PublicSubdomainGatewayForm/>
      </div>
      <div className='lh-copy charcoal'>
        <Trans i18nKey='publicPathGatewayDescription' t={t}>
          <p>Choose which <a className='link blue' href="https://docs.ipfs.tech/concepts/ipfs-gateway/#path" target='_blank' rel='noopener noreferrer'>Path Gateway</a>  you want to use when generating shareable links (Fallback: path gateway used for CIDs that can't be represented in 63 character DNS label)</p>
        </Trans>
        <PublicGatewayForm/>
      </div>
    </Box>
```

##############

Tests: settings page
I've added a couple of tests in ```test/e2e/settings.test.js``` for submit/reset functionality of public gateway

```javascript
test('Submit/Reset Public Subdomain Gateway', async ({ page }) => {
    // Wait for the necessary elements to be available in the DOM
    const publicSubdomainGatewayElement = await page.waitForSelector('#public-subdomain-gateway')
    const publicSubdomainGatewaySubmitButton = await page.waitForSelector('#public-subdomain-gateway-submit-button')
    const publicSubdomainGatewayResetButton = await page.waitForSelector('#public-subdomain-gateway-reset-button')

    // Check that submitting a wrong Subdomain Gateway triggers a red outline
    await submitGatewayAndCheck(page, publicSubdomainGatewayElement, publicSubdomainGatewaySubmitButton, DEFAULT_PATH_GATEWAY, 'focus-outline-red')

    // Check that submitting a correct Subdomain Gateway triggers a green outline
    await submitGatewayAndCheck(page, publicSubdomainGatewayElement, publicSubdomainGatewaySubmitButton, DEFAULT_SUBDOMAIN_GATEWAY + '/', 'focus-outline-green')

    // Check the Reset button functionality
    await resetGatewayAndCheck(publicSubdomainGatewayResetButton, publicSubdomainGatewayElement, DEFAULT_SUBDOMAIN_GATEWAY)
  })

  test('Submit/Reset Public Path Gateway', async ({ page }) => {
    // Custom timeout for this specific test
    test.setTimeout(32000)

    // Wait for the necessary elements to be available in the DOM
    const publicGatewayElement = await page.waitForSelector('#public-gateway')
    const publicGatewaySubmitButton = await page.waitForSelector('#public-path-gateway-submit-button')
    const publicGatewayResetButton = await page.waitForSelector('#public-path-gateway-reset-button')

    // Check that submitting a wrong Path Gateway triggers a red outline
    await submitGatewayAndCheck(page, publicGatewayElement, publicGatewaySubmitButton, DEFAULT_PATH_GATEWAY + '1999', 'focus-outline-red')

    // Check that submitting a correct Path Gateway triggers a green outline
    await submitGatewayAndCheck(page, publicGatewayElement, publicGatewaySubmitButton, DEFAULT_SUBDOMAIN_GATEWAY, 'focus-outline-green')

    // Check the Reset button functionality
    await resetGatewayAndCheck(publicGatewayResetButton, publicGatewayElement, DEFAULT_PATH_GATEWAY)
  })
```

And another test ```src/lib/files.test.js``` for checking subdomain and path gateway urls 

```javascript
it('should get a subdomain gateway url', async () => {
  const ipfs = {}
  const myCID = CID.parse('QmZTR5bcpQD7cFgTorqxZDYaew1Wqgfbd2ud9QqGPAkK2V')
  const file = {
    cid: myCID,
    name: 'example.txt'
    // type: 'file',
  }
  const files = [file]

  const url = new URL(DEFAULT_SUBDOMAIN_GATEWAY)
  const shareableLink = await getShareableLink(files, DEFAULT_PATH_GATEWAY, DEFAULT_SUBDOMAIN_GATEWAY, ipfs)
  const base32Cid = 'bafybeifffq3aeaymxejo37sn5fyaf7nn7hkfmzwdxyjculx3lw4tyhk7uy'
  const rightShareableLink = `${url.protocol}//${base32Cid}.ipfs.${url.host}`
  // expect(res).toBe('https://bafybeifffq3aeaymxejo37sn5fyaf7nn7hkfmzwdxyjculx3lw4tyhk7uy.ipfs.dweb.link')
  expect(shareableLink).toBe(rightShareableLink)
})

it('should get a path gateway url', async () => {
  const ipfs = {}
  // very long CID v1 (using sha3-512)
  const veryLongCidv1 = 'bagaaifcavabu6fzheerrmtxbbwv7jjhc3kaldmm7lbnvfopyrthcvod4m6ygpj3unrcggkzhvcwv5wnhc5ufkgzlsji7agnmofovc2g4a3ui7ja'
  const myCID = CID.parse(veryLongCidv1)
  const file = {
    cid: myCID,
    name: 'example.txt'
  }
  const files = [file]

  const res = await getShareableLink(files, DEFAULT_PATH_GATEWAY, DEFAULT_SUBDOMAIN_GATEWAY, ipfs)
  // expect(res).toBe('https://ipfs.io/ipfs/bagaaifcavabu6fzheerrmtxbbwv7jjhc3kaldmm7lbnvfopyrthcvod4m6ygpj3unrcggkzhvcwv5wnhc5ufkgzlsji7agnmofovc2g4a3ui7ja')
  expect(res).toBe(DEFAULT_PATH_GATEWAY + '/ipfs/' + veryLongCidv1)
})
```

```bash
 git show --stat
commit cdde8884d22d8683afa97e2b0959287fd6769463 (HEAD -> fix-issue-2244, origin/fix-issue-2244)
Author: acul71 
Date:   Tue Sep 3 02:09:47 2024 +0200

    Fix issue #2244: Updated public gateway settings, added subdomain gateway support, and refactored tests.

 public/locales/en/app.json                                                 |   3 +++
 public/locales/en/settings.json                                            |   3 ++-
 src/bundles/files/actions.js                                               |   4 ++-
 src/bundles/gateway.js                                                     | 111 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++---
 src/components/public-gateway-form/PublicGatewayForm.js                    |  10 ++++---
 src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js |  94 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/lib/files.js                                                           |  31 ++++++++++++++++-----
 src/lib/files.test.js                                                      |  38 +++++++++++++++++++++++++-
 src/settings/SettingsPage.js                                               |  23 ++++++++++------
 test/e2e/settings.test.js                                                  |  83 +++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 10 files changed, 374 insertions(+), 26 deletions(-)

```

### CODE REVIEW

File: src/settings/SettingsPage.js [hunk 1]
```diff
@@ -63,15 +64,21 @@ export const SettingsPage = ({
...
+        <div className='lh-copy charcoal'>
         <Title>{t('app:terms.publicGateway')}</Title>
-        <Trans i18nKey='publicGatewayDescription' t={t}>
-          <p>Choose which <a className='link blue' href="http://docs.ipfs.tech/concepts/ipfs-gateway/#public-gateways" target='_blank' rel='noopener noreferrer'>public gateway</a> you want to use to open your files.</p>
-        </Trans>
-        <PublicGatewayForm/>
-      </div>
-    </Box>
+          <Trans i18nKey='publicSubdomainGatewayDescription' t={t}>
+            <p>Choose which <a className='link blue' href="https://docs.ipfs.tech/concepts/ipfs-gateway/#subdomain" target='_blank' rel='noopener noreferrer'>Subdomain Gateway</a> you want to use when generating shareable links (Default: Subdomain Isolation)</p>
```
Comment (lidel) [inline src/settings/SettingsPage.js:71]: nit: this is GUI app, let's keep it short and avoid technical jargon. mentioning origin isolation can be skipped, we link to relevant docs already. 

mind updating here and in `public/locales/en`?

```suggestion
            <p>Select a default <a className='link blue' href='https://docs.ipfs.tech/concepts/ipfs-gateway/#subdomain' target='_blank' rel='noopener noreferrer'>Subdomain Gateway</a> for generating shareable links.</p>
```

File: src/settings/SettingsPage.js [hunk 2]
```diff
@@ -63,15 +64,21 @@ export const SettingsPage = ({
...
+          </Trans>
+          <PublicSubdomainGatewayForm/>
+        </div>
+        <div className='lh-copy charcoal'>
+          <Trans i18nKey='publicPathGatewayDescription' t={t}>
+            <p>Choose which <a className='link blue' href="https://docs.ipfs.tech/concepts/ipfs-gateway/#path" target='_blank' rel='noopener noreferrer'>Path Gateway</a>  you want to use when generating shareable links (Fallback: path gateway used for CIDs that can't be represented in 63 character DNS label)</p>
```
Comment (lidel) [inline src/settings/SettingsPage.js:77]: nit: prefer simpler, single sentence. 

mind updating here and in `public/locales/en`?

```suggestion
            <p>Select a fallback <a className='link blue' href='https://docs.ipfs.tech/concepts/ipfs-gateway/#path' target='_blank' rel='noopener noreferrer'>Path Gateway</a> for generating shareable links for CIDs that exceed the 63-character DNS limit.</p>
```

**Review by lidel — CHANGES_REQUESTED** [review]
Thank you @acul71, this is very useful improvement that allows users to avoid hitting HTTP redirect on page load.

Did  a quick first-ass review, small asks below.

File: src/bundles/gateway.js [hunk 3]
```diff
@@ -76,6 +85,89 @@ const checkImgSrcPromise = (imgUrl) => {
...
+      reject(new Error(`Error when attempting to load img from '${img.src}`))
+    }
+    img.onload = () => {
+      clearTimeout(timer)
+      console.log(`Successfully loaded img from '${img.src}'`)
+      resolve(undefined)
+    }
+    img.src = imgUrl.toString()
+  })
+}
```
[the remark this answers is not in the document]
Comment (acul71) [reply src/bundles/gateway.js]: I've wrote those functions in a separate file than I decided to merge the two files, I was unsure if mods would bother the original composer.
But no I've implemented the nit

File: test/e2e/settings.test.js [hunk 4]
```diff
@@ -25,6 +25,87 @@ test.describe('Settings screen', () => {
...
+   * Function to reset a gateway and verify the reset.
+   * @param {ElementHandle} resetButton - The reset button element to click.
+   * @param {ElementHandle} inputElement - The input element to check.
+   * @param {string} expectedValue - The expected value after reset.
+   */
+  async function resetGatewayAndCheck (resetButton, inputElement, expectedValue) {
+    await resetButton.click()
+    const gatewayText = await inputElement.evaluate(element => element.value)
+    expect(gatewayText).toContain(expectedValue)
+  }
```
Comment (SgtPooki) [inline test/e2e/settings.test.js:72]: nit: would prefer these functions outside the test blocks

File: src/settings/SettingsPage.js [hunk 5]
```diff
@@ -63,15 +64,21 @@ export const SettingsPage = ({
         </div>
       </Box> }
 
-    <Box className='mb3 pa4-l pa2'>
-      <div className='lh-copy charcoal'>
+      <Box className='mb3 pa4-l pa2'>
+        <div className='lh-copy charcoal'>
```
Comment (SgtPooki) [inline src/settings/SettingsPage.js:68]: the indentation level here should be where the opening brace on line 55 is.

```suggestion
    <Box className='mb3 pa4-l pa2'>
      <div className='lh-copy charcoal'>
```

but also, we should have lint/style tools handle this automatically for us

File: src/lib/files.test.js [hunk 6]
```diff
@@ -249,3 +251,37 @@ it('drop multiple directories', async () => {
...
   expectRightOutput(output, expected)
 })
+
+it('should get a subdomain gateway url', async () => {
+  const ipfs = {}
+  const myCID = CID.parse('QmZTR5bcpQD7cFgTorqxZDYaew1Wqgfbd2ud9QqGPAkK2V')
+  const file = {
+    cid: myCID,
+    name: 'example.txt'
+    // type: 'file',
```
Comment (SgtPooki) [inline src/lib/files.test.js:261]:
```suggestion
```

File: src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import { connect } from 'redux-bundler-react'
+import { withTranslation } from 'react-i18next'
+import Button from '../button/Button.js'
+import { checkValidHttpUrl, checkSubdomainGateway, DEFAULT_SUBDOMAIN_GATEWAY } from '../../bundles/gateway.js'
+
+const PublicSubdomainGatewayForm = ({ t, doUpdatePublicSubdomainGateway, publicSubdomainGateway }) => {
+  const [value, setValue] = useState(publicSubdomainGateway)
+  const initialIsValidGatewayUrl = !checkValidHttpUrl(value)
+  const [showFailState, setShowFailState] = useState(initialIsValidGatewayUrl)
+  const [isValidGatewayUrl, setIsValidGatewayUrl] = useState(initialIsValidGatewayUrl)
```
Comment (SgtPooki) [inline src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js:11]: can we merge these into one state variable?

```suggestion
  const [isValidGatewayUrl, setIsValidGatewayUrl] = useState(initialIsValidGatewayUrl)
```

File: src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+  // Updates the border of the input to indicate validity
+  useEffect(() => {
+    setShowFailState(!isValidGatewayUrl)
+  }, [isValidGatewayUrl])
```
Comment (SgtPooki) [inline src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js:16]:
```suggestion
```

File: src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+  // Updates the border of the input to indicate validity
+  useEffect(() => {
+    const isValid = checkSubdomainGateway(value)
+    setIsValidGatewayUrl(isValid)
+    setShowFailState(!isValid)
```
Comment (SgtPooki) [inline src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js:22]:
```suggestion
```

File: src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+  const onSubmit = async (event) => {
+    event.preventDefault()
+
+    let isValid = false
+    try {
+      isValid = await checkSubdomainGateway(value)
+      setShowFailState(!isValid)
+    } catch (e) {
+      setShowFailState(true)
```
Comment (SgtPooki) [inline src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js:35]:
```suggestion
      setIsValidGatewayUrl(true)
    } catch (e) {
      setIsValidGatewayUrl(false)
```

File: src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  return (
+    <form onSubmit={onSubmit}>
+      <input
+        id='public-subdomain-gateway'
+        aria-label={t('terms.publicSubdomainGateway')}
+        placeholder={t('publicSubdomainGatewayForm.placeholder')}
+        type='text'
+        className={`w-100 lh-copy monospace f5 pl1 pv1 mb2 charcoal input-reset ba b--black-20 br1 ${showFailState ? 'focus-outline-red b--red-muted' : 'focus-outline-green b--green-muted'}`}
```
Comment (SgtPooki) [inline src/components/public-subdomain-gateway-form/PublicSubdomainGatewayForm.js:61]:
```suggestion
        className={`w-100 lh-copy monospace f5 pl1 pv1 mb2 charcoal input-reset ba b--black-20 br1 ${!isValidGatewayUrl ? 'focus-outline-red b--red-muted' : 'focus-outline-green b--green-muted'}`}
```

File: test/e2e/settings.test.js (hunk 4 above)
Comment (acul71) [reply test/e2e/settings.test.js]: nit: moved kind helper functions outside the test blocks
```javascript
import { readFile } from 'node:fs/promises'
import { test, expect } from './setup/coverage.js'
import { DEFAULT_PATH_GATEWAY, DEFAULT_SUBDOMAIN_GATEWAY } from '../../src/bundles/gateway.js'

const languageFilePromise = readFile('./src/lib/languages.json', 'utf8')

let languages
const getLanguages = async () => {
  if (languages != null) return languages
  languages = JSON.parse(await languageFilePromise)
  return languages
}

/**
 * Function to check if an element contains a specific class within a maximum wait time.
 * @param {Page} page - The page object.
 * @param {ElementHandle} element - The element to check.
 * @param {string} className - The class name to check for.
 * @param {number} maxWaitTime - Maximum wait time in milliseconds.
 * @param {number} pollInterval - Interval between polls in milliseconds.
 * @returns {Promise<boolean>} - A promise that resolves to a boolean indicating if the class was found.
 */
async function checkClassWithTimeout (page, element, className, maxWaitTime = 16000, pollInterval = 500) {
  const startTime = Date.now()
  while ((Date.now() - startTime) < maxWaitTime) {
    const hasClass = await element.evaluate((el, className) => el.classList.contains(className), className)
    if (hasClass) return true
    await page.waitForTimeout(pollInterval)
  }
  return false
}

/**
 * Function to submit a gateway and check for success/failure.
 * @param {Page} page - The page object.
 * @param {ElementHandle} inputElement - The input element to fill.
 * @param {ElementHandle} submitButton - The submit button element to click.
 * @param {string} gatewayURL - The gateway URL to fill.
 * @param {string} expectedClass - The expected class after submission.
 */
async function submitGatewayAndCheck (page, inputElement, submitButton, gatewayURL, expectedClass) {
  await inputElement.fill(gatewayURL)
  await submitButton.click()
  const hasExpectedClass = await checkClassWithTimeout(page, inputElement, expectedClass)
  expect(hasExpectedClass).toBe(true)
}

/**
 * Function to reset a gateway and verify the reset.
 * @param {ElementHandle} resetButton - The reset button element to click.
 * @param {ElementHandle} inputElement - The input element to check.
 * @param {string} expectedValue - The expected value after reset.
 */
async function resetGatewayAndCheck (resetButton, inputElement, expectedValue) {
  await resetButton.click()
  const gatewayText = await inputElement.evaluate(element => element.value)
  expect(gatewayText).toContain(expectedValue)
}

test.describe('Settings screen', () => {
  test.beforeEach(async ({ page }) => {
    await page.goto('/#/settings')
  })

  test('should show config of IPFS node', async ({ page }) => {
    await page.waitForSelector('text=Addresses')
    await page.waitForSelector('text=Bootstrap')
    await page.waitForSelector('text=PeerID')
    // check PeerID in config to confirm it comes from expected instance
    const id = process.env.IPFS_RPC_ID
    await page.waitForSelector(`text=${id}`)
  })

  test('Submit/Reset Public Subdomain Gateway', async ({ page }) => {
    // Wait for the necessary elements to be available in the DOM
    const publicSubdomainGatewayElement = await page.waitForSelector('#public-subdomain-gateway')
    const publicSubdomainGatewaySubmitButton = await page.waitForSelector('#public-subdomain-gateway-submit-button')
    const publicSubdomainGatewayResetButton = await page.waitForSelector('#public-subdomain-gateway-reset-button')

    // Check that submitting a wrong Subdomain Gateway triggers a red outline
    await submitGatewayAndCheck(page, publicSubdomainGatewayElement, publicSubdomainGatewaySubmitButton, DEFAULT_PATH_GATEWAY, 'focus-outline-red')

    // Check that submitting a correct Subdomain Gateway triggers a green outline
    await submitGatewayAndCheck(page, publicSubdomainGatewayElement, publicSubdomainGatewaySubmitButton, DEFAULT_SUBDOMAIN_GATEWAY + '/', 'focus-outline-green')

    // Check the Reset button functionality
    await resetGatewayAndCheck(publicSubdomainGatewayResetButton, publicSubdomainGatewayElement, DEFAULT_SUBDOMAIN_GATEWAY)
  })

  test('Submit/Reset Public Path Gateway', async ({ page }) => {
    // Custom timeout for this specific test
    test.setTimeout(32000)

    // Wait for the necessary elements to be available in the DOM
    const publicGatewayElement = await page.waitForSelector('#public-gateway')
    const publicGatewaySubmitButton = await page.waitForSelector('#public-path-gateway-submit-button')
    const publicGatewayResetButton = await page.waitForSelector('#public-path-gateway-reset-button')

    // Check that submitting a wrong Path Gateway triggers a red outline
    await submitGatewayAndCheck(page, publicGatewayElement, publicGatewaySubmitButton, DEFAULT_PATH_GATEWAY + '1999', 'focus-outline-red')

    // Check that submitting a correct Path Gateway triggers a green outline
    await submitGatewayAndCheck(page, publicGatewayElement, publicGatewaySubmitButton, DEFAULT_SUBDOMAIN_GATEWAY, 'focus-outline-green')

    // Check the Reset button functionality
    await resetGatewayAndCheck(publicGatewayResetButton, publicGatewayElement, DEFAULT_PATH_GATEWAY)
  })

  test('Language selector', async ({ page }) => {
    const languages = await getLanguages()
    for (const lang of Object.values(languages).map((lang) => lang.locale)) {
      // click the 'change language' button
      const changeLanguageBtn = await page.waitForSelector('.e2e-languageSelector-changeBtn')
      await changeLanguageBtn.click()

      // wait for the language modal to appear
      await page.waitForSelector('.e2e-languageModal')

      // create a promise that resolves when the request for the new translation file is made
      const requestForNewTranslationFiles = page.waitForRequest((request) => {
        if (lang === 'en') {
          // english is the fallback language and we can't guarantee the request wasn't already made, so we resolve for 'en' on any request

          return true
        }
        const url = request.url()

        return url.includes(`locales/${lang}`) && url.includes('.json')
      })

      // select the language
      const languageModalButton = await page.waitForSelector(`.e2e-languageModal-lang_${lang}`)
      await languageModalButton.click()

      // wait for the language modal to disappear
      await page.waitForSelector('.e2e-languageModal', { state: 'hidden' })
      await requestForNewTranslationFiles

      // check that the language has changed
      await page.waitForSelector('.e2e-languageSelector-current', { text: languages[lang].nativeName })

      // confirm the localStorage setting was applied
      const i18nLang = await page.evaluate('localStorage.getItem(\'i18nextLng\')')
      expect(i18nLang).toBe(lang)
    }
  })
})

```

File: src/lib/files.test.js [hunk 12]
```diff
@@ -1,5 +1,7 @@
 /* global it, expect */
-import { normalizeFiles } from './files.js'
+import { normalizeFiles, getShareableLink } from './files'
+import { DEFAULT_SUBDOMAIN_GATEWAY, DEFAULT_PATH_GATEWAY } from '../bundles/gateway'
```
Comment (SgtPooki) [inline src/lib/files.test.js:3]:
```suggestion
import { normalizeFiles, getShareableLink } from './files.js'
import { DEFAULT_SUBDOMAIN_GATEWAY, DEFAULT_PATH_GATEWAY } from '../bundles/gateway.js'
```

File: src/bundles/gateway.js [hunk 13]
```diff
@@ -1,20 +1,29 @@
 import { readSetting, writeSetting } from './local-storage.js'
 
-export const DEFAULT_GATEWAY = 'https://ipfs.io' // TODO: switch to dweb.link when https://github.com/ipfs/kubo/issues/7318
+// export const DEFAULT_GATEWAY = 'https://ipfs.io' // TODO: switch to dweb.link when https://github.com/ipfs/kubo/issues/7318
+export const DEFAULT_PATH_GATEWAY = 'https://ipfs.io'
```
Comment (SgtPooki) [inline src/bundles/gateway.js:4]:
```suggestion
// TODO: switch to dweb.link when https://github.com/ipfs/kubo/issues/7318
export const DEFAULT_PATH_GATEWAY = 'https://ipfs.io'
```

**Review by SgtPooki — APPROVED** [review]
lgtm, two minor changes requested. will wait to merge for @lidel to take a peek

**Comment by acul71** [comment]
@lidel 
Hello, I requested a review is the code ok? Thanks, Bye

**Review by SgtPooki — APPROVED** [review]
lgtm, works locally..

**Review by lidel — APPROVED** [review]
Thank you @acul71!

Pushed small fix in https://github.com/ipfs/ipfs-webui/pull/2255/commits/44a43be7d40f01aa3685032cbcb52db5331e0a13 to stop spamming console + avoid hitting dweb.link by all users who run with default gateway.

**Pull request merged**